bootloader/src/pyi_utils.c

```diff
@@ -94,6 +94,40 @@ static int argc_pyi = 0;
 static void process_apple_events_new();
 #endif
 
+char *
+pyi_strjoin(const char *first, const char *sep, const char *second){
+    /* join first and second string, using sep as separator.
+     * any of them may be either a zero-terminated string or NULL.
+     * sep will be only used if first and second string are not empty.
+     */
+    int first_len=0, sep_len=0, second_len=0;
+    char *result;
+    if (first) {
+        first_len = strlen(first);
+    }
+    if (sep) {
+        sep_len = strlen(sep);
+    }
+    if (second) {
+        second_len = strlen(second);
+    }
+    result = malloc(first_len + sep_len + second_len + 1);
+    if (!result) {
+        return NULL;
+    }
+    *result = '\0';
+    if (first_len) {
+        strcat(result, first);
```

Comment:
Can we use snprintf instead?
As long as value doesn't contain literal zero byte, it's got much simpler and safer API.